### PR TITLE
Ssl mutual authentication implementation

### PR DESCRIPTION
**Pull request by asimarslan**
Client SSL/TLS mutual authentication support implementation and tests

### CODE REVIEW

**Comment by asimarslan** [comment]
verify

**Comment by asimarslan** [comment]
run-net

**Comment by asimarslan** [comment]
verify

File: Hazelcast.Net/Hazelcast.Config/SSLConfig.cs [hunk 1]
```diff
@@ -93,6 +121,43 @@ internal string GetCertificateName()
...
+        {
+            return GetProperty(CertificatePassword);
+        }
+
+        internal SslProtocols GetSslProtocol()
+        {
+            var sslProtocol = GetProperty(SslProtocol);
+            if (sslProtocol == null)
+            {
+#if NET40
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Config/SSLConfig.cs]: `NET40`  represents the .net framework and includes all versions.
There is also NETSTANDART2_0 for net core side.
We need a compile time decision here as `None` option is added on Net core only.

File: Hazelcast.Net/Hazelcast.Config/SSLConfig.cs [hunk 2]
```diff
@@ -93,6 +121,43 @@ internal string GetCertificateName()
...
+                return SslProtocols.Tls;
+#else
+                return SslProtocols.None;
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Config/SSLConfig.cs]: Yes supported but deprecated. But None parameter means best supported TLS/SSL version to be chosen by OS level.

File: Hazelcast.Test/Hazelcast.Client.Test/ClientSslBaseTest.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            RemoteController.startMember(cluster.Id);
+            var clientConfig = new ClientConfig();
+            clientConfig.GetNetworkConfig().AddAddress("localhost:5701");
+            clientConfig.GetNetworkConfig().SetSSLConfig(CreateSslConfig(isSslEnabled, validateCertificateChain,
+                validateCertificateName, checkCertificateRevocation, certSubjectName, clientCertificate, certPassword));
+            clientConfig.GetGroupConfig().SetName(cluster.Id).SetPassword(cluster.Id);
+            Client = HazelcastClient.NewHazelcastClient(clientConfig);
+        }
+
+        private static SSLConfig CreateSslConfig(bool isSslEnabled, bool? validateCertificateChain, bool? validateCertificateName,
```
Comment (sancar) [inline Hazelcast.Test/Hazelcast.Client.Test/ClientSslBaseTest.cs]: `bool?`
What is the purpose of making a bool Nullable?

Comment (asimarslan) [reply Hazelcast.Test/Hazelcast.Client.Test/ClientSslBaseTest.cs]: null represents default values of the related configuration

File: Hazelcast.Net/Hazelcast.Client.Connection/ClientConnection.cs [hunk 4]
```diff
@@ -111,14 +113,45 @@ internal sealed class ClientConnection
...
+                            Logger.Finest(string.Format("Cannot load client certificate:{0}.", cerPath));
+                            throw;
+                        }
+                        
+                        var enabledSslProtocols = sslConfig.GetSslProtocol();
+                        var checkCertificateRevocation = sslConfig.IsCheckCertificateRevocation();
+                        sslStream.AuthenticateAsClient(certificateName, clientCertificates, enabledSslProtocols, checkCertificateRevocation);
+                        if (!sslStream.IsMutuallyAuthenticated && clientCertificates.Count > 0)
+                        {
+                            throw new CryptographicException("Mutual authentication failed.");
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Connection/ClientConnection.cs]: I will add a test

File: Hazelcast.Net/Hazelcast.Config/SSLConfig.cs [hunk 5]
```diff
@@ -23,10 +26,30 @@ namespace Hazelcast.Config
     public class SSLConfig
     {
         /// <summary>
-        /// Certifate Name; CN part of the Certificate Subject.
+        /// Certificate Name; CN part of the Certificate Subject.
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Config/SSLConfig.cs]: This is builtin supported by .NET. I'll fix the doc

**Pull request merged**